subxt/src/backend/legacy/mod.rs

```diff
@@ -84,183 +85,249 @@ impl<T: Config + Send + Sync + 'static> Backend<T> for LegacyBackend<T> {
-                let hash = h.hash();
-                (h, BlockRef::from_hash(hash))
+        let methods = self.methods.clone();
+
+        let retry_sub = retry_stream(move || {
+            let methods = methods.clone();
+            Box::pin(async move {
+                let sub = methods.chain_subscribe_all_heads().await?;
+                let sub = sub.map(|r| {
+                    r.map(|h| {
+                        let hash = h.hash();
+                        (h, BlockRef::from_hash(hash))
+                    })
+                });
+                Ok(StreamOf(Box::pin(sub)))
             })
-        });
-        Ok(StreamOf(Box::pin(sub)))
+        })
+        .await?;
+
+        Ok(retry_sub)
     }
 
     async fn stream_best_block_headers(
         &self,
     ) -> Result<StreamOfResults<(T::Header, BlockRef<T::Hash>)>, Error> {
-        let sub = self.methods.chain_subscribe_new_heads().await?;
-        let sub = sub.map(|r| {
-            r.map(|h| {
-                let hash = h.hash();
-                (h, BlockRef::from_hash(hash))
+        let methods = self.methods.clone();
+
+        let retry_sub = retry_stream(move || {
+            let methods = methods.clone();
+            Box::pin(async move {
+                let sub = methods.chain_subscribe_new_heads().await?;
+                let sub = sub.map(|r| {
+                    r.map(|h| {
+                        let hash = h.hash();
+                        (h, BlockRef::from_hash(hash))
+                    })
+                });
+                Ok(StreamOf(Box::pin(sub)))
             })
-        });
-        Ok(StreamOf(Box::pin(sub)))
+        })
+        .await?;
+
+        Ok(retry_sub)
     }
 
     async fn stream_finalized_block_headers(
         &self,
     ) -> Result<StreamOfResults<(T::Header, BlockRef<T::Hash>)>, Error> {
-        let sub: super::rpc::RpcSubscription<<T as Config>::Header> =
-            self.methods.chain_subscribe_finalized_heads().await?;
-
-        // Get the last finalized block immediately so that the stream will emit every finalized block after this.
-        let last_finalized_block_ref = self.latest_finalized_block_ref().await?;
-        let last_finalized_block_num = self
-            .block_header(last_finalized_block_ref.hash())
-            .await?
-            .map(|h| h.number().into());
-
-        // Fill in any missing blocks, because the backend may not emit every finalized block; just the latest ones which
-        // are finalized each time.
-        let sub = subscribe_to_block_headers_filling_in_gaps(
-            self.methods.clone(),
-            sub,
-            last_finalized_block_num,
-        );
-        let sub = sub.map(|r| {
-            r.map(|h| {
-                let hash = h.hash();
-                (h, BlockRef::from_hash(hash))
+        let this = LegacyBackend {
+            methods: self.methods.clone(),
+            storage_page_size: self.storage_page_size,
+        };
```

Comment:
Is there any reason not to just do `let this = self.clone();` here and below?

Reply:
It's a workaround/hack because T: Clone is not satisfied here and it won't work

Reply:
I could implement manually though if you want

Reply:
Ooh yeah lets manually impl `Clone` then because it's a bit useless with that default impl as it stands. whoops!